### PR TITLE
Move opentracing-javascript to model where immutable span context is supported.

### PR DESCRIPTION
**Pull request by oibe**
fixes https://github.com/opentracing/opentracing-javascript/issues/45

@bcronin @bensigelman @yurishkuro

### CODE REVIEW

File: src/span_context.js [hunk 1]
```diff
@@ -11,55 +11,21 @@
...
-     * @param  {string} key
-     *         The key for the given trace attribute.
-     * @return {string}
-     *         String value for the given key, or undefined if the key does not
-     *         correspond to a set trace attribute.
+     * @return {Object} - returns the new SpanContext with key:value pair set
+     * as baggage on the new context.
      */
-    getBaggageItem(key) {
+    withBaggageItem(key, value)  {
```
Comment (yurishkuro) [inline src/span_context.js]: withBaggageItem is actually an implementation detail. Does it need to be in the API?  From [this](https://github.com/opentracing/opentracing-javascript/blob/master/src/span.js#L34) it looks like the API method `context()` is always retrieving the current context from the `impl` and wrapping it, so `withBaggageItem` can be handled completely internally to the span implementation.

Comment (oibe) [inline src/span_context.js]: Ok.  On another note I see that opentracing-go has forEachBaggageItem on its spanContext.  Do we need that here?

Comment (yurishkuro) [inline src/span_context.js]: I would leave it off for another PR. The iterator style may not be idiomatic for JS.

File: test/imp/noop_imp.js [hunk 2]
```diff
@@ -1,8 +1,6 @@
 function NoopSpanContextImp() {
 };
-NoopSpanContextImp.prototype.setBaggageItem = function (key, value) {
-};
-NoopSpanContextImp.prototype.getBaggageItem = function (key) {
+NoopSpanContextImp.prototype.withBaggageItem = function (key, value) {
```
Comment (bensigelman) [inline test/imp/noop_imp.js]: I don't think we need this...

File: src/span_context.js [hunk 3]
```diff
@@ -11,59 +11,6 @@
  * <trace_id, span_id, sampled> tuple).
  */
 export default class SpanContext {
```
Comment (bensigelman) [inline src/span_context.js]: I can see an argument for specifically pointing to inject() and extract() in this class comment... since SpanContext no longer has any (!) methods in `ot-javascript`, it's now just a "concept" and a little hard to comprehend without the context of inject and extract.

**Comment by bensigelman** [comment]
/me feels nervous having an interface that has no methods. I would like some guidance from .js natives about the most idiomatic way to represent iteration over baggage items.

I'm fine merging this in the interim, though it would be good to wait for word from @bcronin assuming you can wait until Monday (he is on a short vacay).

**Comment by oibe** [comment]
I don't mind waiting till monday to hear bcronon's opinion.

File: src/span_context.js (hunk 3 above)
Comment (bcronin) [inline src/span_context.js]: For what it's worth, I'm fine with `SpanContext` having no methods. I'd like to do something this week to make the implementation code more inheritance-based than purely implicit per #35 and I think the "empty" class will fit in better then (and if it *doesn't* I can remove it then).

**Comment by bcronin** [comment]
This looks fine to me :)

**Comment by yurishkuro** [comment]
@bcronin @bensigelman are you guys ok to merge this?

**Comment by bcronin** [comment]
Looks good to me
